### ISSUE TITLE
Map types are not generated properly in Go

### ISSUE THREAD

**Issue by mohabusama** [Jumpscale/go-raml#27]
Hi,

Assuming the following input RAML1.0 file:

```YAML
#%RAML 1.0
title: UserApi
version: 0.1
baseUri: /users
types:
 user:
    properties:
        username:
            type: string
        emails:
          properties:
            []:
              type: string
    example:
        username: bob
        emails:
            work: bob@company.com
            home: bob@example.com
/users:
  post:
    description: Create a new user
    body:
        application/json:
            type: user
```

the generated Go server model for the `User` is

```go
package main

type User struct {
        Emails   string `json:"emails"`
        Username string `json:"username"`
}
```

`Emails` field is invalid, since it is expected to be a `map`.

Ref: http://docs.raml.org/specs/1.0/#raml-10-spec-map-types

**Comment by iwanbk** [comment]
@mohabusama 
It seems i missed this http://docs.raml.org/specs/1.0/#raml-10-spec-map-types spec.

For now, you can use `string{}` to specify a map, you can see the spec here http://docs.raml.org/specs/1.0/#raml-10-spec-type-expressions

About your example, i think it is not valid syntax because:
- a property can't have property
- http://docs.raml.org/specs/1.0/#raml-10-spec-map-types is about map Type, not map property.
So, the example
```
#%RAML 1.0 
title: My API With Types
types:
  MapOfNumbers:
    type: object
    properties:
      []:
        type: number
```
Will be translated to `type MapOfNumber map[string]float64`

CMIIW

**Comment by iwanbk** [comment]
@mohabusama 
http://docs.raml.org/specs/1.0/#raml-10-spec-map-types has been partially implemented in #28 
check https://github.com/Jumpscale/go-raml/pull/28#issuecomment-188125443 for the details

As for your use case, you can use `string{}` as defined in http://docs.raml.org/specs/1.0/#raml-10-spec-type-expressions

**Comment by mohabusama** [comment]
Hi @iwanbk 

Do you use specific tool for syntax validation?

Generating with `string{}` works as expected indeed. 
However, I am using [API Workbench](http://apiworkbench.com/) and for some reason it complains about:

```YAML
emails: 
    type: string{}
```
with `Syntax error:Expected "|" or end of input but "{" found`. However it doesn't complain about `string[]` or about the `properties` way in my original spec.

Which makes me wonder if this is considered valid (assuming it is part of the User `properties`):

```YAML
emails:
    type: object
    properties:
        []:
            type: string
```
Generating with go-raml it produces:

```go
package main

type User struct {
        Emails   object `json:"emails"`
        Username string `json:"username"`
}
```
I guess `Emails object` could be another bug in the generation, no?

I have added a comment on #28.

Thanks,
Mohab

**Comment by iwanbk** [comment]
```
Do you use specific tool for syntax validation?
```
No

You can find about `string{}` in http://docs.raml.org/specs/1.0/#raml-10-spec-type-expressions

```
emails:
    type: object
    properties:
        []:
            type: string
```
It is valid RAML, but unfortunately still no support for it.
Properties type is `map[string]Property` 
In yaml parser we use, `[]` is not valid `string` and `interfaces{}`

```
I guess Emails object could be another bug in the generation, no?
```
I need to see the RAML file of this.

**Comment by robvanmieghem** [comment]
I actually like the `string{}` syntactic sugar notation a lot better then 
```
[]:
            type: string
```
It is indeed too bad that `string{}` is not supported by API workbench but that is a bug on their side.

**Comment by robvanmieghem** [comment]
Was looking for a way to implement this way of declaring maps since the api-console for example does not support the `type{}` way of declaring maps. The problem is that 
```
properties:
    []:
      type: string
```
is not accepted by the yaml parser as valid. I don't really think this is valid yaml 1.2 either so I created a bug on the raml project for this: https://github.com/raml-org/raml-spec/issues/286

**Comment by robvanmieghem** [comment]
ok, so according to the response on the ticket at the raml spec:
```
properties:
    "[]":
      type: string
```
is better. I tested this syntax with the api-console and this works fine. @iwanbk I don't think this would be hard to do now since the yaml parser accepts the key now.

**Comment by iwanbk** [comment]
implemented in #35
